rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,192 @@
+
+We will introduce a new token type, `oidc-jwt`, and add an additional field to the Token resource to allow the issuer URL to be specified (`issuer_url`).
+
+Registration flow:
+
+1. Client is configured by the user to use `oidc-jwt` joining with a specific provider. The client then uses the provider-specific logic to obtain a token.
+2. The client will call the `RegisterUsingToken` endpoint, providing the OIDC JWT token that it has collected, and specifying the name of the Teleport provisioning token which should be used to verify it.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have allow-listed (RS[256, 384, 512])
+5. The server will check the `kid` of the JWT header, and obtain the relevant JWK from the cache or from the specified issuers well-known JWKS endpoint. It will then use the JWK to validate the token has been signed by the issuer.
+6. Other key claims of the JWT will be validated:
+  - Ensure the Issued At Time (iat) is in the past.
+  - Ensure the Expiry Time (exp) is in the future.
+7. The user's [configured Common Expression Language rule](#configuration) for the token will be evaluated against the claims, to ensure that the token is allowed to register with the Teleport cluster.
+8. Certificates will be generated for the client. In the case of bot certificates, they will be treated as non-renewable, to match the behaviour of IAM registration for bot certificates.
+
+We will re-use `go-jose@v2` for validation of JWTs, since this library is already in use within Teleport.
+
+#### Caching JWKs
+
+Special attention should be given to the logic around caching the JWKs.
+
+We will cache these for two reasons:
+
+- Improve the performance of validating JWTs, as we will not need to make a HTTPS request to the issuer.
+- Improve the reliability, as we can validate JWTs even if the issuer is experiencing some downtime.
+- Reduce the impact of Teleport on an issuer. If onboarding a large number of nodes, we do not want to unduly place pressure on the issuer. 
+
+We should keep in mind the following considerations:
+
+- When we are presented with a JWT with a previously unseen `kid`, we should re-check the issuer's JWKs, as they may have begun issuing tokens with a new JWK.
+- We should ensure that the TTL of the cache entries is relatively short, as we want to allow an issuer to revoke a JWK that has been stolen.
+
+We will implement this cache in-memory, as the data set is relatively small and it's cheap for us to repopulate this after a service restart.
+
+#### Configuration
+
+We will leverage the existing Token configuration kind as used by static tokens and IAM joining:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: my-gce-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: [Node]
+  join_method: oidc-jwt
+  issuer_url: https://accounts.google.com
+  allow:
+  - aud: "noah.teleport.sh"
```

Comment:
Unfortunately, proto doesn't really play well with nested maps, which we will need here. I'll need to think on this a little bit more. I'm open to suggestions on how we can handle this, potentially with a custom json encoder/decoder for a proto type that looks like:

```proto
message NestedStringMap {
  map<string, NestedStringMap> children = 1;
  map<string, string> data = 2;
}
```

Reply:
I had a quick look at CEL (I'm not completely opposed to it FWIW). Would we need to come up with some custom nested map type to implement that as well?

We already do have a predicate language we use for search and certain conditions in roles based on https://github.com/vulcand/predicate, it might be suitable, looks like we support matching json, have you had a look at https://github.com/gravitational/teleport/blob/fa1dab196a037e9d8a0d912fbef40ed75fde1ae2/lib/services/parser.go#L399?

I would imagine a final token config like this, which I think would be feasible to implement

```yaml
kind: token
version: v2
metadata:
  name: my-gce-token
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: [Node]
  join_method: oidc-jwt
  issuer_url: https://accounts.google.com
  token_matcher: >
    equals(token.aud, "noah.teleport.sh") &&
    equals(token.google.compute_engine.project_id, "my-project") &&
    contains(token.google.compute_engine.instance_name, "teleport")
```

Reply:
Using CEL or the predicate language, we'd simply take the user's inputted rules as a string, so we wouldn't have the issue with encoding it in Proto like we would with the "simple" YAML matching.

My concerns with our predicate language, is that I'm not sure that our current codebase (predicate.Contains, predicate.Equals etc) supports anything other than map[string]string. We'd need to adapt it to support a map containing strings, numbers and nested maps in order to be able to properly process the structure of incoming claims. CEL would theoretically support that out of the box, but I'm hesitant to introduce another configuration language to Teleport.

Reply:
There is definitely some support for JSON, which should be perfect for matching the JWT. On my cluster I can run a search like `resource.spec.hostname == "one-auth" && resource.spec.rotation.phase == "standby"`

Reply:
Odd, I'm potentially misunderstanding the code here. I'll do a scratchpad to test it tomorrow.

Reply:
May need some work to make `contains` work, and I'm not sure about lists, but it may be doable

Reply:
Ah interesting, yeah I am not really familiar with the details on this, but to me it seems like a suitable predicate language that we already use in a couple places